File: backend/src/models/Document.js
import mongoose from 'mongoose';

const documentSchema = mongoose.Schema({
    user: {
        type: mongoose.Schema.Types.ObjectId,
        required: true, // Legacy field, kept for safety
        ref: 'User',
    },
    uploadedBy: {
        type: mongoose.Schema.Types.ObjectId,
        required: true,
        ref: 'User',
    },
    name: {
        type: String,
        required: true,
    },
    type: {
        type: String,
        required: true,
    },
    size: {
        type: String,
        required: true,
    },
    url: {
        type: String,
        required: true,
    },
    folder: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Folder',
        default: null,
    },
    team: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Team',
        required: true
    },
    isDownloadable: {
        type: Boolean,
        default: true   // All roles can download by default; head/admin can restrict
    },
    cloudinaryId: {
        type: String,
        default: null   // Cloudinary public_id; null for legacy local files
    },
}, {
    timestamps: true,
});

const Document = mongoose.model('Document', documentSchema);

export default Document;

File: backend/models/document.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Document struct {
	ID             primitive.ObjectID  `bson:"_id,omitempty" json:"_id"`
	User           primitive.ObjectID  `bson:"user" json:"user"`
	UploadedBy     primitive.ObjectID  `bson:"uploadedBy" json:"uploadedBy"`
	Name           string              `bson:"name" json:"name"`
	Type           string              `bson:"type" json:"type"`
	Size           string              `bson:"size" json:"size"`
	URL            string              `bson:"url" json:"url"`
	Folder         *primitive.ObjectID `bson:"folder,omitempty" json:"folder"`
	Team           primitive.ObjectID  `bson:"team" json:"team"`
	IsDownloadable bool                `bson:"isDownloadable" json:"isDownloadable"`
	CloudinaryID   *string             `bson:"cloudinaryId,omitempty" json:"cloudinaryId"`
	CreatedAt      time.Time           `bson:"createdAt,omitempty" json:"createdAt,omitempty"`
	UpdatedAt      time.Time           `bson:"updatedAt,omitempty" json:"updatedAt,omitempty"`

	// Populated fields
	UploadedByDetails *UserSummary `bson:"-" json:"uploadedByDetails,omitempty"`
}
